Pr/subtree1

hunger opened the pull request:
There are some things to clean up still -- but I'd like to do those later:

* I would like to use "subtree" for the DynamicNode in the ItemTree more consistently.
* I am not particularly happy with `Component::subtree_index()`. Maybe rename that to `Component::index_property()` or something?
* I would like to remove the visitor from the ComponentVTable and replace that with rust code walking the logical ItemTree
* We need to add a function to the Component to overrride the default focus chain. I left that out for now as we do not support that on the slint side. The focus code looks at the Component ItemTree only, so it is easy to add extra information to override the Component-wide focus chain once we have that.
* I'd like to change the Component's parent_item function to "parent_subtree" or something and make that return the Component/Node-index of the place where this Component is in the parent Component only. The parent_item is readily available from the ItemTree and does not need to be handled by generated code at all.

The `focus_event_propagation_2" test currently fails for me. Looking into that right now.

ogoffart commented on internal/core/model.rs:
```diff
@@ -968,7 +974,21 @@ impl<C: RepeatedComponent> Repeater<C> {
     }
 
-    /// Return true if the Repeater is empty
+    /// Return the amount of item currently in the component
```
i guess this comment is not accurate.

I was wondering why you used a range in the vtable instead of just the size,  until i saw the implementation.

hunger replied:
We could discuss what we want to do about that support for offset we have in some places, but that is outside of this patch I think:-)

ogoffart commented on internal/core/component.rs:
```diff
@@ -32,17 +43,35 @@ pub struct ComponentVTable {
     /// And the return value is passed by &mut because ItemWeak has a destructor
     pub get_item_tree: extern "C" fn(core::pin::Pin<VRef<ComponentVTable>>) -> Slice<ItemTreeNode>,
 
+    // FIXME: This does return an invalid ItemWeak now that points to the parent repeater!
```
Maybe we should just rename this to `parent_node`.  Maybe also rename Item{Rc,Weak} to something like Node{Rc,Weak}.    or ItemNodeRc ?

hunger replied:
I do not want to rename ItemRc and co.: Those are really nice to have at the logical (Component-spanning) Item tree!

The Component is below that level though: It only covers the physical Itemtree with all the repeaters and stuff... at this level we will end up with thigs that are either an Item or a Subtree. It would be cool if we could keep those two levels separate on the type system level.

ogoffart commented on internal/core/component.rs:
```diff
@@ -32,17 +43,35 @@ pub struct ComponentVTable {
     pub get_item_tree: extern "C" fn(core::pin::Pin<VRef<ComponentVTable>>) -> Slice<ItemTreeNode>,
 
+    // FIXME: This does return an invalid ItemWeak now that points to the parent repeater!
+    // FIXME: Get rid of the index and make this always return the "Item" that connects this component
+    //        to its parent-component?
```
i don't really understand what this is about.

hunger replied:
I would like this to end up as somehting along this line:

`fn parent_subtree(ComponentRc) -> Option<(ComponentRc, usize)>,` (in FFI-friendly;-)

and then make this unconditionally return the ItemWeak we currently return when the index is `0`.

There is no need to duplicate the code to find the parent item when staying inside the components item tree in all the different generated code we have! So this is the only interesting bit of information this method provides right now.

ogoffart reviewed (commented):
Looks good!  just left a couple of comments

hunger commented:
The failed check seems unrelated to my changes:-)

hunger commented:
Only the last change is new.

ogoffart commented on internal/compiler/widgets/fluent/std-widgets.slint:
```diff
@@ -308,6 +308,10 @@ export TabImpl := Rectangle {
 
+    border-radius: 2px;
+    border-width: fs.has-focus ? 3px : 0px;
+    border-color: Palette.themeSecondary;
```
Did you look at the fluent design from Microsoft for that?

hunger replied:
No, I just did the easiest thing possible to figure out where the focus went :-)

ogoffart replied:
this is what i used for the sesign IIRC.  https://developer.microsoft.com/en-us/fluentui#/controls/web/pivot
sadly there is no focus in there.
How does it look with that change?

ogoffart commented on internal/core/component.rs:
```diff
@@ -82,6 +82,9 @@ pub struct ComponentVTable {
 }
 
+#[cfg(test)]
+pub(crate) use ComponentVTable_static;
```
why is this like this?

hunger replied:
This is the way to use a macro defined in a different module of the same crate, introduced in rust 1.32 AFAICT.

ogoffart reviewed (commented):
looking forward to see this merged.

ogoffart commented on tests/cases/focus/keyboard_focus.slint:
```diff
@@ -0,0 +1,102 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
```
This test is not really testing anythingnew, is it?.  It just will add time to the test without increasing the test coverage.

You should add some ```rust or cpp comments that sends keystroke to it. 
If it is meant to be a manual test only, move it to tests/manual.
otherwise you can also leave it out.

hunger replied:
I'll move it to manual for now. The navigation itself should be covered in the rust code anyway.

hunger commented on internal/core/items.rs:
```diff
@@ -162,6 +162,73 @@ pub struct ItemVTable {
     ) -> RenderingResult,
 }
 
+fn find_sibling_outside_repeater(
```
I do not see many users for this but the two places it is used in here, so that's where I put them.

How about extracting itemrc + its helper cfunctions and classes from here into itemrc.rs? That would then contain the "wrapper code" plus the tree navigation, leavint items.rs for the actual implmentation.

hunger commented on internal/core/items.rs:
```diff
@@ -234,8 +301,142 @@ impl ItemRc {
+    }
+}
+
+impl Eq for ItemRc {}
```
True!

hunger commented:
I derived the ItemRc's Eq trait and dropped the change to the fluent Tab Widget and moved the test case.

I do want to implement some tests for focus handling -- since that will also include that the widgets used in the test are well-behaved, but did not do that yet.

tronical commented on api/cpp/include/slint.h:
```diff
@@ -846,6 +848,16 @@ class Repeater
         return { &C::static_vtable, const_cast<C *>(&(**x.ptr)) };
     }
 
+    void component_at(int i, vtable::VWeak<private_api::ComponentVTable> *result) const
```
I suggest to return the VWeak by value here. After all this is called by generated code, right?

It would also be consistent with the Repeater in Rust.

hunger replied:
Good point, done.

tronical commented on internal/compiler/generator/cpp.rs:
```diff
@@ -1321,6 +1383,25 @@ fn generate_sub_component(
+        subtrees_components_cases.push(format!(
+            "\n        case {i}: {{
+                {e_u}
+                self->{id}.component_at(subtree_index, result);
```
And I guess here it would be `*result =`

tronical commented on internal/compiler/generator/rust.rs:
```diff
@@ -796,6 +838,33 @@ fn generate_sub_component(
+                };
+            }
+
+            fn index_property(self: ::core::pin::Pin<&Self>) -> usize {
```
It looks to me that this function has only one call site. Maybe it should be inlined there?

hunger replied:
I would like to postpone this: I do not trust my rust-generator magic enough to not break something unrelated when I do that:-)

tronical commented on internal/core/window.rs:
```diff
@@ -359,6 +372,76 @@ impl Window {
+
+    fn move_focus(self: Rc<Self>, start_item: ItemRc, forward: impl Fn(ItemRc) -> ItemRc) {
+        let mut current_item = start_item;
+        let mut visited = vec![];
```
I think an `extern crate alloc;` may fix this

tronical commented on internal/core/item_tree.rs:
```diff
@@ -102,6 +104,82 @@ impl ItemTreeNode {
+
+impl<'a> ComponentItemTree<'a> {
+    /// Create a new `ItemTree` from its raw data.
+    pub fn new(item_tree: &'a [ItemTreeNode]) -> Self {
```
All of the call sites pass `comp_ref_pin.as_ref().get_item_tree().as_slice()`. It think the function should take a component ref pin as argument.

tronical commented on internal/core/window.rs:
```diff
@@ -359,6 +372,76 @@ impl Window {
+    /// Take the focus_item out of this Window
+    ///
+    /// This sends the FocusOut event!
+    fn take_focus_item(self: Rc<Self>) -> Option<ItemRc> {
```
Since this function doesn’t seem to need to consume or clone the Rc, I think it can either take a &self or &Rc<Self>. Then you can save the extra clone call at the call site(s).

hunger replied:
Good point, going to change this.

tronical commented on internal/core/window.rs:
```diff
@@ -359,6 +372,76 @@ impl Window {
+    pub fn focus_next_item(self: Rc<Self>) {
+        let component = self.component();
+        let start_item = self
+            .clone()
```
Example of clone call that seems not necessary:)

tronical reviewed (commented):
Some comments, mostly nitpicks :). So feel free to ignore. Looks like the right direction!!

hunger commented:
Deriving Eq for ItemRc does not work due to the Component in the struct there:-/
